graphql_compiler/macros/macro_edge/validation.py

```diff
@@ -61,6 +65,55 @@ def _validate_macro_ast_with_macro_directives(schema, ast, macro_directives):
                 .format(directive_definition.name, len(macro_data)))
 
 
+def _validate_macro_ast_directives(ast, inside_fold_scope=False):
+    """Check that the macro is using non-macro direcives properly.
+
+    Restrictions on use of directives:
+    - @output and @output_source are disallowed
+    - @macro_edge_target is not allowed to be inside a @fold scope
+
+    Args:
+        ast: GraphQL AST describing a subtree of the macro
+        inside_fold_scope: bool, whether the subtree is within a @fold scope
+    """
+    subselection_inside_fold_scope = inside_fold_scope
+    directives_with_no_restrictions = (
+        FilterDirective.name,
+        TagDirective.name,
+        OptionalDirective.name,
+        RecurseDirective.name,
+        MacroEdgeDefinitionDirective.name,
+    )
+    disallowed_directives = (
+        OutputDirective.name,
+        OutputSourceDirective.name,
+    )
+    for directive in ast.directives:
+        name = directive.name.value
+        if name in directives_with_no_restrictions:
+            pass
+        elif name in disallowed_directives:
+            raise GraphQLInvalidMacroError(u'Macros are not allowed to use the {} directive. '
+                                           u'Found usage {}'.format(name, directive))
+        elif name == FoldDirective.name:
+            subselection_inside_fold_scope = True
+        elif name == MacroEdgeTargetDirective.name:
+            if inside_fold_scope:
+                raise GraphQLInvalidMacroError(
+                    u'The @macro_edge_target cannot be inside a fold scope.')
+        else:
+            raise AssertionError(u'Unexpected directive name found: {} {}'.
+                                 format(directive.name.value, directive))
```

Comment:
nit: Very interesting that this works. I would have put the `.` together with `format` though, because you see that the syntax highlighting is a bit messed up at the moment (it thinks `format` is a global).